### DISCUSSION TITLE
Facing initialisation delays when using SINGLE_FILE flag

### DISCUSSION THREAD

**Discussion by bansal21ankit** [emscripten-core/emscripten#19746] [General]
Hi Team,

**_Intro :_** I am using [ZXing-CPP's WASM](https://github.com/zxing-cpp/zxing-cpp/tree/master/wrappers/wasm) wrapper in one of my project. It uses Emscripten + CMake to generate the .js and .wasm files. Due to some limitation my project doesn't support a separate wasm file due to which I had to add the SINGLE_FILE flag in CMake flags list.

**_Issue :_** This workaround works fine and solves the limitation but introduces almost 50-60 sec delay during the initialisation where getBinaryPromise() method in the generated js file makes a fetch() command on the wasmBinaryFile variable which is (after using SINGLE_FILE) a base64 string of almost 1.2 million characters.
```
function getBinaryPromise(binaryFile) {
            if (!wasmBinary && (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER)) {
                if (typeof fetch == 'function') {
                    return fetch(binaryFile, { credentials: 'same-origin' })
                        .then(function (response) {
                            if (!response['ok']) {
                                throw "failed to load wasm binary file at '" + binaryFile + "'";
                            }
                            return response['arrayBuffer']();
                        })
                        .catch(function () {
                            return getBinary(binaryFile);
                        });
                }
            }
            return Promise.resolve().then(function () {
                return getBinary(binaryFile);
            });
        }
```

**_Workaround :_** Now, to fix this bug for the time-being I am modifying the code in generated file to just this:
```
function getBinaryPromise(binaryFile) {
          return Promise.resolve().then(function () { return getBinary(binaryFile); });
        }
```

**_Query :_** With this post I just wanted to seek advice from other experts around and ask if there is any other graceful way to fix this so that I do not need to modify the generated files ?

**Comment by sbc100** [comment]
If I understand you correctly you are saying the fetch promise takes 50-60s seconds to resolve, but the getBinary() promise resolves much quick, but that they both achieve the same end result?

I wonder why the fetch method takes so long in comparison?

**Reply by bansal21ankit** [reply]
Correct, when I debugged the call made due to this fetch command got halted but once that was completed the following steps were quicker.

**Reply by sbc100** [reply]
Did the fetch promise succeed or end up in the catch() block?

**Reply by bansal21ankit** [reply]
Just debugged again, it is landing in success after a halt of 37 sec.
<img width="978" alt="Screenshot 2023-06-29 at 11 17 34 PM" src="https://github.com/emscripten-core/emscripten/assets/5683899/2a651aa5-701d-48a3-a0bf-ce0080736833">

**Reply by bansal21ankit** [reply]
Hi @sbc100,
Is it possible to not have that fetch call in the first place, for SINGLE_FILE scenarios since we have the data available as binary anyway ?

**Comment by sbc100** [comment]
What version of emscripten are you using?

**Reply by bansal21ankit** [reply]
```
EmscriptenSDK % em++ --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.35 (a90c5abb489c516f59c0d16864d953d4bedd37da)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by bansal21ankit** [comment]
Any suggestions would be appreciated.
Please help guys.

**Comment by bansal21ankit** [comment]
Adding a message to get focus of community on this ticker.

**Comment by bansal21ankit** [comment]
Hi Guys,
I am still struggling with this issue if anyone can look into it ?

**Reply by sbc100** [reply]
Yes I believe this should have been fixed in #19789, which was released in 3.1.43.   Have you tried using 3.1.43?

**Reply by bansal21ankit** [reply]
Code Changes wise looks like it should work but I'll verify and update too.
Thanks for the fixes.
